src/core/memory.cc

```diff
@@ -1687,6 +1698,13 @@ void *allocate_slowpath(size_t size) {
     } else {
         ptr = allocate_large(size, should_sample);
     }
+
+    if (!ptr && fallback_to_system_nest_count && original_malloc_func) {
+        // we are in a fallback scope, call the system malloc
+        alloc_stats::increment(alloc_stats::types::fallback_allocs);
+        ptr = original_malloc_func(size);
+    }
```

Comment:
From a perf prospective would it be better to have something like:

```
struct scoped_use_system_alloc { ... }
struct scoped_throw_on_over_alloc { ... }

try {
  scoped_throw_on_over_alloc scoped;
  return do_protobuf_stuff();
} catch (const std::bad_alloc& e) {
  std::ignore = e;
}
scoped_use_system_alloc scoped;
return do_protobuf_stuff();
```

Reply:
That way we don't have this check always and we can fold it into the existing check of if we're on a reactor thread at the beginning?

Reply:
Which check are you referring to? The highlighted check is in a the slow path of the allocator and is very fast since !ptr is a simple check which is almost always true, so I doubt it will make any measurable difference, though I should add `unlikely` here.

I do have performance tests for the allocator which I can run to double check.

Reply:
So to clarify, the use case I imagined is just:

```
scoped_use_system_alloc scoped;
return do_protobuf_stuff();
```

No need for `disable_abort_on_alloc_failure_temporarily`.

Reply:
Well for the fallback yes you wouldn't need it. I was thinking there would be a scoped way to just entirely disable the seastar allocator (as opposed to doing a fallback). I don't really feel strongly, if this is already on the slow path then I'm good with it.

Reply:
In principle it could make "zero" difference since we already have a check for allocation failure, so we could put it inside there which means it never affects the non-failure path at all, but due to the existing structure of the code that is a bit awkward (e.g., the use of `finish_allocation` which is shared among > 1 code paths). _Maybe_ the compiler is smart enough to build a path which only checks `ptr` once out of the inlined result of all this.

Reply:
> I was thinking there would be a scoped way to just entirely disable the seastar allocator (as opposed to doing a fallback). I don't really feel strongly, if this is already on the slow path then I'm good with it.

This is also possible, but it seems like a bad idea? My view is: as long as there is memory, we want to use the seastar allocator. We shouldn't opt out protobug stuff from the seastar allocator globally. 

Rather, in the specific case that an allocation would _otherwise_ fail in a scoped region fall back to the system allocator.

There are a lot of reasons to use the seastar allocator as much as possible (lazy to list them but will if you prod me :)).

Reply:
No I agree, it's just where the fallback happens. If we have this slow path in the allocator, then it's fine, but I was thinking of not touching the allocator at all - I know it's a very critical hotpath - but to push the fallback to the application by requiring them to opt into throwing on bad allocs first, then falling back manually, since there should be very very few places we actually do this it would be better to move it up and not take this hit on all allocations, but since there is a cold path to stick this check into I'm fine with your approach, they are at the end of the day the same, it's just where the check happens.

Reply:
>  but to push the fallback to the application by requiring them to opt into throwing on bad allocs first, then falling back manually, since there should be very very few places we actually do this it would be better to move it up and not take this hit on all allocations, 

That's also appealing, but I can't see how you'd do it? In your example in the application fallback you use `scoped_use_system_alloc` : that has to be implemented in the allocator, and in fact it's more expensive because I think this requires a hot-path check? Asking because I may be missing something here.

By doing it as a fallback inside the allocator you get the chance to move the check to the place which is only executed on the slow/failure path (in this case "slow" not just "failure" due to code structure.

Reply:
> that has to be implemented in the allocator (I think), and in fact it's more expensive because I think this requires a hot-path check

It does have to be implemented in the allocator, and we already have a check in the hotpath for this: https://github.com/redpanda-data/seastar/blob/398be413afad0b6fca48b5d781c9bd9bf10033ff/src/core/memory.cc#L1711

But I guess you're right that it's simpler to just add this to the slowpath.

Reply:
I added benchmark results at: https://github.com/redpanda-data/seastar/pull/209#issuecomment-2867847605

Reply:
@rockwotj closing this as resolved, please feel free to re-open if you disagree.